Pseudo transient continuation to speed up (#25944)
idaholab/moose: framework/doc/content/source/postprocessors/PseudoTimestep.md

# PseudoTimestep

> The methods implemented in this postprocessor compute timesteps adaptively using Pseudo Transient Continuation strategies. This allows for steady state problems solved via time marching approcah to reach the final state rapidly.

Reviewer: ```suggestion
The methods implemented in this postprocessor compute timesteps adaptively using Pseudo Transient Continuation strategies. This allows for steady state problems solved via time marching approach to reach the final state rapidly.
```

Author: The .md documentation is to be extended, and I'd of course run a spell checker at the end, but as a draft, it felt overkill.

As a reminder the time marching approach to computing steady states reframes a problem in discrete form $M\frac{\partial \mathbf u}{\Delta t}=F(\mathbf u)$ as

\begin{equation}
M\frac{\mathbf u_{n+1}-\mathbf u_n}{\Delta t}=F(\mathbf u_{n+1})
\end{equation}

## Overview

This object computes a timestep to accelerate the convergence to steay state using a pseudo transient.

> The change in timestep is determined by the residual behaviour from one iteration to another, i.e. small changes in  residual indicate larger timesteps are allowed, while large changes indicate a decrease is necessary.

Reviewer: Comment on the "iterations_window" parameter.

> There main methods are recognized and implemented.

Reviewer: ```suggestion
The main methods are recognized and implemented.
```
is this correct?  The wording doesn't seem right.  I would just delete this sentence.

Author: Same as above, this is a draft to be extended and run a spell checker before the review process.

Following [!citep](bucker2009cfl), we recognize and implement three main methods.

The user must make a method choice, between `SER`, `EXP` and `RDM`, which are implemented as listed below. 
All methods require a parameter `alpha`, which controls how sensitive the timestep should be with respect to residual changes, and `initial_dt` to provide a first timestep length. If nothing is known about the problem we recommend `initial_dt = 1` and `alpha=2`, keeping in mind that a high `alpha` corresponds to a higher sensitivity to residual changes. More specific choices for fluid dynamics problems are available in [!citep](bucker2009cfl) or [!citep](ceze2013pseudo). 

Methods supported

- Switched evolution relaxation (SER), as exemplified in [fv_burgers_SER.i](test/tests/postprocessors/pseudotimestep/fv_burgers_SER.i),

\begin{equation}
\Delta t_k =\Delta t_{k-1} \cdot \bigg(\frac{R_{k-\ell}}{R_{k}}\bigg)^{\alpha}
\end{equation}

> where $\alpha$ is a user chosen parameter. The l2-norm of the steady-state residual at step $k$ is $R_k$, and the residual at $\ell$ iterations before is denoted as $R_{k-\ell}$.

Author: Of course always > 0, I'll add this

- Exponential progression (EXP), as exemplified in [fv_burgers_EXP.i](test/tests/postprocessors/pseudotimestep/fv_burgers_EXP.i),

\begin{equation}
\Delta t_k =\Delta t_{0} \cdot \alpha^k
\end{equation}

> where $\alpha$ is a user chosen parameter, $k$ is the current iteration step.

Author: I reserve all these recommendations for the literature cause it's really problem dependent or one can just set a random alpha and see how it decays.. but I'll add one more sentence.

- Residual Difference Method (RDM), as exemplified in [fv_burgers_RDM.i](test/tests/postprocessors/pseudotimestep/fv_burgers_RDM.i),

\begin{equation}
\Delta t_k =\Delta t_{k-1} \cdot \alpha^{\frac{R_{k-1}-R_k}{R_{k-1}}}
\end{equation}

The pseudotimestepper block 
!listing test/tests/postprocessors/pseudotimestep/fv_burgers_RDM.i
    block=Postprocessors/pseudotimestep

> As noted also in [!citep](bucker2009cfl) the EXP method has an infinite growth, so for this method a `max_dt` parameter may be recommended. If no `max_dt` is provided by the user then infinite growth of the timestep is not bounded and the user will be informed by a message at the console. Ideally this method is used in conjunction with a steady state detection, i.e. setting `steady_state_detection = true` and a `steady_state_tolerance`.

Author: I had already, this says outdated.

\begin{equation}
\Delta t_k =\Delta t_{k-1} \cdot \bigg(\frac{R_{k-\ell}}{R_{k}}\bigg)^{\alpha}
\end{equation}

where $\alpha$ is a user chosen parameter. The l2-norm of the steady-state residual at step $k$ is $R_k$, and the residual at $\ell$ iterations before is denoted as $R_{k-\ell}$. To set a number of previous iterations corresponding to $\ell$ the user can prescribe an integer value for the parameter [!param](/Postprocessors/PseudoTimestep/iterations_window).

- Residual Difference Method (RDM), as exemplified in [fv_burgers_RDM.i](test/tests/postprocessors/pseudotimestep/fv_burgers_RDM.i),

\begin{equation}
\Delta t_k =\Delta t_{k-1} \cdot \alpha^{\frac{R_{k-1}-R_k}{R_{k-1}}}
\end{equation}

This implementation is the `RDM` method variant as found in [!citep](ceze2013pseudo), other variants are available in e.g. [!citep](bucker2009cfl).

- Exponential progression (EXP), as exemplified in [fv_burgers_EXP.i](test/tests/postprocessors/pseudotimestep/fv_burgers_EXP.i),

\begin{equation}
\Delta t_k =\Delta t_{0} \cdot \alpha^k
\end{equation}

where $\alpha$ is a user chosen parameter, $k$ is the current iteration step.

> As noted also in [!citep](bucker2009cfl) the EXP method has an infinite growth, so for this method a [!param](/Postprocessors/PseudoTimestep/max_dt) parameter may be recommended. If no [!param](/Postprocessors/PseudoTimestep/max_dt) is provided by the user then infinite growth of the timestep is not bounded and the user will be informed by a message at the console. Ideally this method is used in conjunction with a steady state detection, i.e. setting [!param](/Postprocessors/Executioner/steady_state_detection) to `true` and a [!param](/Postprocessors/Executioner/steady_state_tolerance).

Reviewer 2: ```suggestion
As noted also in [!citep](bucker2009cfl) the EXP method has an infinite growth, so for this method a [!param](/Postprocessors/PseudoTimestep/max_dt) parameter may be recommended. If no [!param](/Postprocessors/PseudoTimestep/max_dt) is provided by the user then infinite growth of the timestep is not bounded and the user will be informed by a message at the console. Ideally this method is used in conjunction with a steady state detection, i.e. setting [!param](/Executioners/Transient/steady_state_detection) to `true` and a [!param](/Executioners/Transient/steady_state_tolerance).
```

As a reminder, the time marching approach to computing steady states reframes a problem in discrete form $M\frac{\partial \mathbf u}{\Delta t}=F(\mathbf u)$, where $M$ is the mass matrix and $F()$ an operator gathering the right-hand side discrete terms.
Using implicit Euler, the problem reads

\begin{equation}
M\frac{\mathbf u_{n+1}-\mathbf u_n}{\Delta t}=F(\mathbf u_{n+1})
\end{equation}

To be consistent with the literature, the current implementation has been tested only on implicit Euler.

## Overview

This object computes a timestep to accelerate the convergence to the steady state using pseudo-transient continuation.
The change in timestep is determined by the residual behavior from one iteration to another, i.e., small changes in residual indicate larger timesteps are allowed. In contrast, significant changes in the residual indicate a timestep decrease is necessary.
Following [!citep](bucker2009cfl), we recognize and implement three main methods.

The user must make a method choice, between `SER`, `EXP` and `RDM`, which are implemented as listed below. 

> All methods require a parameter `alpha`, which controls how sensitive the timestep should be with respect to residual changes, and `initial_dt` to provide a first timestep length. If nothing is known about the problem we recommend `initial_dt = 1` and `alpha=2`, keeping in mind that a high `alpha` corresponds to a higher sensitivity to residual changes. More specific choices for fluid dynamics problems are available in [!citep](bucker2009cfl) or [!citep](ceze2013pseudo). 

Author: Fixed
